### ISSUE TITLE
Allow deploy and undeploy tasks to occur simultaneously for different channels

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3237]
Right now if a single channel is hanging during a deploy, it will block all subsequent channels in the request from being deployed. This is a big concern on normal server startup when all channels are initially deployed. Donkey should do something similar to what we're doing in ChannelStatusServlet (essentially submit multiple deploys to an executor). Right now the private deployChannel method in DEC is synchronized, but it probably doesn't need to be (nor does it even make sense right now since it's only ever called in the public deployChannels method which is already synchronized).

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3334
Reporter: narupley
Created: 2014-06-11T07:58:01.000-0700

**Comment by rbeckman-nextgen** [comment]
Note that the current behavior is intentional to guarantee integrity of the channel cache -and also to support multiple servers on the same database-. Any changes made to this area requires substantial testing.

Imported Comment. Original Details:
Author: wayneh
Created: 2014-06-11T10:12:06.000-0700

**Comment by rbeckman-nextgen** [comment]
The **private* deployChannel method doesn't need to be synchronized, since it is only called from the *public** deployChannels method. The public deploy channels method would _remain_ synchronized, so that only a single request should happen at a time. Currently the public deployChannels method just iterates through each channel ID and calls the private method. The changes here would call for the public method to submit multiple tasks to an executor, and those tasks would call the private method simultaneously for different channels. This is similar to the changes made to ChannelStatusServlet for start/stop/etc. tasks. I agree, obviously something like this (as with any issue) is going to require testing.

However I'm not sure that the synchronization itself has anything to do with multiple servers on the same database. Different servers are going to be running in completely different JVMs, so both servers are still going to be able to get into their respective deployChannels methods at the same time. Each server will only allow one deploy request, but multiple deploy requests can still be occurring simultaneously. If the goal was to only allow one deploy request across multiple servers, it would require some sort of centralized or peer-to-peer locking system, like the clustering extension provides. That said, it's still "better" restricting requests to one per server than allowing more.

Imported Comment. Original Details:
Author: narupley
Created: 2014-06-27T08:19:25.000-0700

**Comment by rbeckman-nextgen** [comment]
You're right that this change wasn't made to support multiple servers. It was for MIRTH-2756 and MIRTH-2693. We'll have to verify those use cases after any changes we make for this issue.

Note that if we leave deployChannels synchronized and take it off from deployChannel, one channel hanging will not block the rest of the channels from deploying/starting. However as long as that channel is hanging at the start then it will still prevent all future deploy/undeploy requests. I think in 99% of cases the channel gets stuck starting/stopping, not deploy/undeploying. In fact I'm not sure I've ever seen an instance of a channel stuck at deploying/undeploying. The problem is that on a deploy we also start the channel from the same thread. On an undeploy we first stop the channel from the same thread (which leads to the reverse issue MIRTH-3316). I think you can see where I'm going here. We should split up so start does not happen directly in a deploy thread and stop does not happen directly in the undeploy thread.

Any channels should deploy as stopped, and then a separate start request should be made automatically by the server. This means that the deploy can still be synchronized but should hang much much less. Then each channel will be able to start even if another one is hanging. For undeploy, a separate stop request should be made first on all the channels before ever calling the synchronized undeploychannels. Note that we'll have to make sure that when a channel is being stopped for an undeploy it can't be started again.

Imported Comment. Original Details:
Author: wayneh
Created: 2014-06-27T10:46:55.000-0700

**Comment by rbeckman-nextgen** [comment]
We have seen channels hang on deploy (or at least I certainly have). In fact it happened recentishly to a client because a deploy script was hanging (technically a global deploy script, but it can just as easily be a channel deploy script doing the hanging). It doesn't even have to be JavaScript either; it's possible for a channel to hang on deploy if it does any Velocity replacement because of MIRTH-3136 (though that's rare and dependent on high-volume / race conditions).

Since the deploy script is user-defined/arbitrary, it's just as susceptible to hanging as a channel is during a stop task (where it can hang for a variety of reasons like bad transformer scripts).

Imported Comment. Original Details:
Author: narupley
Created: 2014-06-27T17:40:14.000-0700

**Comment by rbeckman-nextgen** [comment]
So it seems like each channel basically needs its own queue (maybe executor) to performs deploy/undeploy operations in the order they were sent to the server. The code that adds to the queue should be synchronized but that's really it. This is slightly different then the solution you mentioned above because that one would still allow the same channel to be deployed/undeployed simultaneously if one task gets stuck and then a different executor thread picks up a task for the same channel. This is very similar to how each Channel has executor for the tasks within the channel, so maybe DonkeyEngineController just needs something like that for deploy/undeploy.

This would theoretically mean that no channel would ever block another channel, only itself and it should also still guarantee that MIRTH-2756 isn't a problem.

Imported Comment. Original Details:
Author: wayneh
Created: 2014-06-27T17:57:34.000-0700

**Comment by rbeckman-nextgen** [comment]
If we managed to remove the synchronization across all channels and only guaranteed that a single channel could be be deploying/undeploying simultaneously, how would we deal with the global deploy and shutdown scripts? Right now, while one deploy job is executing the global deploy script is run once at the start. Even if another deploy job comes in with different channels, the second job has to wait till the first one is completed before it can even run the global deploy script and start deploying channels. If we are saying we want to allow other channels to deploy simultaneously, we cannot block so the global deploy script would have to run again, possibly before the first job has finished.

We could just create an executor specifically for global deploy and shutdown scripts, so whenever a job comes in we just queue it up and let fate decide what happens. As long as it is not running multiple times simultaneously they will probably get the result they want. The same would have to go for shutdown script. The global deploy and shutdown scripts would probably use the same executor so they can't run simultaneously.

Speaking of deploy and shutdown scripts, when a redeploy all is executed all channels will be undeployed and the global shutdown script will run. Then the global deploy script will run before all channels are deployed. This is expected. When one or more channels are deployed (but not redeploy all), they are first undeployed, then the global deploy script runs, then they are all deployed. I'm not sure if this is correct. Should the global shutdown script run after the channels are undeployed (assuming any were in fact undeployed).

Imported Comment. Original Details:
Author: wayneh
Created: 2014-07-29T16:52:06.000-0700

**Comment by rbeckman-nextgen** [comment]
I would think that the global shutdown script should be running if the global deploy script runs afterward, but perhaps there is some justification as to why it was done this way.  What does 2.x do?

Imported Comment. Original Details:
Author: jacobb
Created: 2014-07-29T17:01:44.000-0700

**Comment by rbeckman-nextgen** [comment]
2.x works more like we would expect. The global shutdown scripts will run if any channels get undeployed on a deploy task. So that's probably its own issue.

Imported Comment. Original Details:
Author: wayneh
Created: 2014-07-29T18:56:01.000-0700

**Comment by rbeckman-nextgen** [comment]
I think there's a couple of wrong things happening regarding the global scripts here. When deploying 1+ channel, the global deploy script is run first, then each channel is undeployed (if its deployed already) and then deployed. What I think should probably happen is: all channels that are selected will be undeployed first, then the global shutdown script should run, then global deploy, then they should be deployed. 

Basically this would be the same format as if a redeploy all happened, just with fewer channels. This would really help in making things more consistent and will make the de-synchronization process easier.

Imported Comment. Original Details:
Author: wayneh
Created: 2014-07-29T19:11:46.000-0700

**Comment by rbeckman-nextgen** [comment]
What about if all the channels you selected haven't yet been deployed? In that case nothing would be undeployed, so is it still correct to execute the global shutdown script? And what about if all the channels you select _are_ currently deployed, but are now disabled? In that case, although those channels would be undeployed, nothing would actually be deployed, so it might not make sense to run the global deploy script. FYI you might think that we would disallow that on the Channels view, but that doesn't appear to be the case:

!deploy1.PNG!

When you deploy those selected disabled channels, the global deploy script still runs even though nothing was deployed... currently deployed channels aren't undeployed either and we show an alert dialog.

Imported Comment. Original Details:
Author: narupley
Created: 2014-07-29T20:07:25.000-0700

**Comment by rbeckman-nextgen** [comment]
Interesting. If one disabled channel is selected the Deploy Channel task will not show up. If multiple disabled channels are selected then it will show up. In 2.x, anytime a channel was actually undeployed, the global shutdown script was run. Anytime Deploy Channel or Redeploy All is pressed, the global deploy script will run even if no channels actually got deployed. For 3.x, I think the shutdown script should work the same way, as long as a channel was undeployed then it should run. I'm not sure how it should work for global deploy. We would think that as long as a channel gets deployed then it should run. If no channels actually get deployed though, I think the global deploy script should still run since that's what people are expecting.

Imported Comment. Original Details:
Author: wayneh
Created: 2014-07-30T09:52:07.000-0700

**Comment by rbeckman-nextgen** [comment]
Multiple channels can now be deployed/undeployed simultaneously. If one channel gets stuck during the start of a deploy or the stop of an undeploy, it will no longer block other channels from being deployed or undeployed. If a channel fails at the stop portion of an undeploy task, it will no longer be undeployed.

Refer to MIRTH-3397 for further discussion about the global deploy and shutdown scripts.

Imported Comment. Original Details:
Author: wayneh
Created: 2014-08-11T20:03:29.000-0700

**Comment by rbeckman-nextgen** [comment]
Verified this problem existed in older builds by creating several test channels. I added an infinite loop to a channel's deploy script and attempted to deploy/undeploy subsequent channels after this channel to no avail. Conversely, I added an infinite loop to a channel's shutdown script and attempted to perform the same operations with no success. I exported both of these channels to 3.1 and verified that these infinite looping scripts no longer block subsequent channels from deploying, redeploying all, or undeploying.

Imported Comment. Original Details:
Author: jaysenp
Created: 2014-08-21T11:10:48.000-0700
